await for async jest config

hisco opened the pull request:
## Summary
Currently jest cli `require` the dynamic config module synchronously, this PR enables loading jest config module async. 

## Test plan
Tested project using `yarn test` - got some issues not related to my change:
```bash
/jest/packages/jest-haste-map/src/lib/FSEventsWatcher.ts
14:1  error  Unused eslint-disable directive (no problems were reported)
/jest/packages/jest-transform/src/ScriptTransformer.ts
578:3  warning  Arrow function expected no return value  consistent-return
594:3  warning  Arrow function expected no return value  consistent-return
```
In addition created a tiny project and tested the async config module using `yarn link` - no issues were observed.

scotthovestadt commented:
You can run tests with `yarn jest`, looks like there are a few failing.

Can you give me an example of how this is used and also update documentation with that as well?

SimenB commented on packages/jest-config/src/__tests__/readConfig.test.ts:
```diff
@@ -3,8 +3,8 @@
 test('readConfig() throws when an object is passed without a file path', () => {
-  expect(() => {
-    readConfig(
+  expect(async () => {
```
Need to await the `expect` and change matcher to be `rejects.toThrowError`

SimenB replied:
goes for a lot od the test changes

hisco replied:
Fixed, thanks!

hisco commented:
@scotthovestadt thanks

- I run `yarn test` the falling tests indeed show up.
- Updated the contribution section.
- Updated the configuration section.

I added an example of such use is in the docs section, here a more advanced section:
```ts
import * as request from 'request-promise';
export async ()=>{
  return await request('http://some-remote')
}
```

Basically for my use cases, I found a way to do it synchronically, I think it would be much  nicer to do it async.

jeysal commented:
+1 for breaking, put it in a major 😅

hisco commented:
@SimenB I didn't notice any integration test for current implementation, do you have an idea how to mock this scenario?

Thanks in advance!

SimenB commented:
This is a really small breaking change, so we can land it in 25. @hisco could you rebase on master?

hisco commented:
@SimenB I did, unfortunately it's completely broken:
Latest commit broke it (`6bd4e560e288b515ebdf8325f6219522f538dbec`)
At commit `8d0b5e2234d378920eb87112933c5c03016fb98e` everything was working properly.

SimenB commented:
@hisco sorry, I missed your answer, then completely forgot about this 😬 

I made `readConfig` async as part of support `jest.config.mjs` in #9431, so _most_ of the changes in this PR are already on master. It should be a matter of invoking the function after loading in `packages/jest-config/src/readConfigFileAndSetRootDir.ts` and adding a test plus updating docs. Are you up for finishing this, by any chance?

SimenB commented:
@hisco ping 🙂

hisco commented:
It seems that most of my changes are already in master I don't think it's a result of a merge with my code so I won't take the credit...
The only thing missing was the docs, please use #10001 instead it's a much cleaner branch.
